### PR TITLE
feat!: implement new Parser-js interface + migrate to TypeScript

### PR DESCRIPTION
**Pull request by smoya**
**Description**

This PR updates the current parser by:

- Migrates the whole project to TypeScript.
- Implement the new Parser-JS 2.0 [Schema Parser Interface](https://github.com/asyncapi/parser-js/tree/next-major#custom-schema-parsers).

**Related issue(s)**
Fixes https://github.com/asyncapi/avro-schema-parser/issues/166

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -23,28 +27,64 @@
...
+    "js-yaml": "^4.1.0",
+    "jsonpath-plus": "^7.2.0",
+    "lodash": "^4.17.21",
+    "node-fetch": "2.6.7",
+    "ramldt2jsonschema": "^1.2.3",
+    "webapi-parser": "^0.5.0"
+  },
   "devDependencies": {
-    "@asyncapi/parser": "^1.17.0",
+    "@asyncapi/parser": "^2.0.0-next-major.6",
```
Comment (smoya) [inline package.json:52]: Still need to update to latest version were some required types are exported.

File: tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+      "baseUrl": "./src",
+      "target": "es6",
+      "module": "esnext",
+      "lib": [
+        "esnext"
+      ],
+      "declaration": true,
+      "allowJs": true,
+      "skipLibCheck": true,
+      "esModuleInterop": true,
```
Comment (dalelane) [inline tsconfig.json:13]: similar question with the config here - was this trial-and-error to get our existing code to work with typescript, or a set of options you've used before and like, or something?

Comment (smoya) [reply tsconfig.json:13]: Copied from Parser-JS 2.0 development https://github.com/asyncapi/parser-js/blob/next-major/tsconfig.json

File: .eslintrc [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+  - eslint:recommended
+  - plugin:@typescript-eslint/eslint-recommended
+  - plugin:@typescript-eslint/recommended
+  - plugin:sonarjs/recommended
+  - plugin:security/recommended
+
+parserOptions:
+  ecmaVersion: 2018
+
+rules:
```
[the remark this answers is not in the document]
Comment (smoya) [reply .eslintrc:24]: Copied from Parser-JS 2.0 development https://github.com/asyncapi/parser-js/blob/next-major/.eslintrc, created by @magicmatatjahu

**Review by dalelane — APPROVED** [review]
LGTM
(and a nudge that I need to get better at TS 😀 )

**Comment by smoya** [comment]
@magicmatatjahu would you mind reviewing this? 🙏

File: package.json [hunk 4]
```diff
@@ -1,14 +1,19 @@
 {
   "name": "@asyncapi/avro-schema-parser",
-  "version": "1.1.0",
+  "version": "2.0.0",
```
Comment (derberg) [inline package.json:3]: bot changes version, please remove this change from here and `package-lock.json`

File: package.json [hunk 5]
```diff
@@ -23,28 +28,64 @@
     "schema",
     "parser"
   ],
-  "author": "Fran Mendez (fmvilas.com)",
```
Comment (derberg) [inline package.json:26]: why did you remove it? Fran was/is the author of this package

File: README.md [hunk 6]
```diff
@@ -2,19 +2,50 @@
...
-```
+```bash
 npm install @asyncapi/avro-schema-parser
+// OR
+yarn add @asyncapi/avro-schema-parser
 ```
 
 ## Usage
 
 ```js
```
Comment (derberg) [inline README.md:43]: we should still keep examples with `require`
not everyone use TS and `imports` in nodejs are native in  latest versions only, afaik

File: package.json (hunk 5 above)
Comment (magicmatatjahu) [reply package.json:26]: We usually add this:

```json
  "author": {
    "name": "The AsyncAPI maintainers",
    "url": "https://www.asyncapi.com"
  },
```

Comment (smoya) [reply package.json:26]: > why did you remove it? Fran was/is the author of this package

Removed it accidentally because of copy-paste party. @derberg  Should I just leave @fmvilas or rather add what @magicmatatjahu suggested?

File: src/index.ts [hunk 7]
```diff
@@ -274,7 +347,6 @@ async function processUnionSchema(jsonSchema, avroDefinition, isTopLevel, record
   return jsonSchema;
 }
 
-module.exports.avroToJsonSchema = async function avroToJsonSchema(avroDefinition) {
-  validateAvroSchema(avroDefinition);
+export async function avroToJsonSchema(avroDefinition: AvroSchema) {
```
[the remark this answers is not in the document]
Comment (smoya) [reply src/index.ts]: It is being tested in isolation in our tests https://github.com/asyncapi/avro-schema-parser/blob/master/tests/to-json-schema.test.js

[the remark this answers is not in the document]
Comment (magicmatatjahu) [reply src/index.ts]: aaa ok 😄 Thanks!

File: test/avro-schema-parser.spec.ts [hunk 8]
```diff
@@ -0,0 +1,441 @@
+import fs from 'fs';
```
[the remark this answers is not in the document]
Comment (smoya) [reply test/avro-schema-parser.spec.ts:1]: Added!

File: README.md (hunk 6 above)
Comment (smoya) [reply README.md:44]: Done!

File: package.json (hunk 5 above)
Comment (derberg) [reply package.json:26]: well, I say leave it as it was originally, as it reflects reality 😄 he created it, and if we really think is is needed, we can add maintainers into `contributors` array. Anyway, not really related to this PR

**Comment by derberg** [comment]
@smoya Sonar mentions 2 code smells

File: package.json [hunk 9]
```diff
@@ -23,28 +28,64 @@
     "schema",
     "parser"
   ],
```
Comment (smoya) [inline package.json:30]:
```suggestion
  ],
  "author": "Fran Mendez (fmvilas.com)",
```

File: package.json (hunk 5 above)
Comment (smoya) [reply package.json:26]: Reverted https://github.com/asyncapi/avro-schema-parser/pull/168/files#r1006720043

**Comment by smoya** [comment]
@derberg smells fixed

File: README.md [hunk 10]
```diff
@@ -40,16 +72,72 @@ channels:
               type: int
 `
 
-await parser.parse(asyncapiWithAvro)
+const { document } = await parser.parse(asyncapiWithAvro);
+```
+
+```js
+const { Parser } = require('@asyncapi/parser');
+const { AvroSchemaParser } = require('@asyncapi/raml-dt-schema-parser');
```
Comment (magicmatatjahu) [inline README.md:80]:
```suggestion
const { AvroSchemaParser } = require('@asyncapi/avro-schema-parser');
```

File: README.md [hunk 11]
```diff
@@ -40,16 +72,72 @@ channels:
...
+        payload:
+          $ref: 'https://schemas.example.com/user'
+`
+
+const { document } = await parser.parse(asyncapiWithAvro);
+```
+
+```js
+const { Parser } = require('@asyncapi/parser');
+const { AvroSchemaParser } = require('@asyncapi/raml-dt-schema-parser');
```
Comment (magicmatatjahu) [inline README.md:137]:
```suggestion
const { AvroSchemaParser } = require('@asyncapi/avro-schema-parser');
```

File: README.md [hunk 12]
```diff
@@ -65,16 +153,41 @@ channels:
...
+        payload:
+          $ref: 'local/path/to/file/user'
+`
+
+const { document } = await parser.parse(asyncapiWithAvro);
+```
+
+```js
+const { Parser } = require('@asyncapi/parser');
+const { AvroSchemaParser } = require('@asyncapi/raml-dt-schema-parser');
```
Comment (magicmatatjahu) [inline README.md:187]:
```suggestion
const { AvroSchemaParser } = require('@asyncapi/avro-schema-parser');
```

File: README.md [hunk 13]
```diff
@@ -125,21 +239,72 @@ channels:
           $ref: 'https://LY422RBU2HN6JQ5T:+f8wz9a0iM06AX7xfwbzSM9YPw/JIkr22Cvl5EKT5Hb1d/nz5nOpbXV/vZC+Iz5c@example.europe-west3.gcp.confluent.cloud/subjects/test/versions/1/schema'
 `
 
-await parser.parse(asyncapiWithAvro)
+const { document } = await parser.parse(asyncapiWithAvro);
+```
+
+```js
+const { Parser } = require('@asyncapi/parser');
+const { AvroSchemaParser } = require('@asyncapi/raml-dt-schema-parser');
```
Comment (magicmatatjahu) [inline README.md:247]:
```suggestion
const { AvroSchemaParser } = require('@asyncapi/avro-schema-parser');
```

File: README.md [hunk 14]
```diff
@@ -125,21 +239,72 @@ channels:
...
+const avroDiagnostics = diagnostics.filter(d => d.code === 'asyncapi-schemas-v2');
+console.log(avroDiagnostics);
+```
 
 ```js
-const assert = require('assert');
-const parser = require('@asyncapi/parser');
-const avroSchemaParser = require('./index');
+const { Parser } = require('@asyncapi/parser');
+const { AvroSchemaParser } = require('@asyncapi/raml-dt-schema-parser');
```
Comment (magicmatatjahu) [inline README.md:304]:
```suggestion
const { AvroSchemaParser } = require('@asyncapi/avro-schema-parser');
```

**Review by magicmatatjahu — COMMENTED** [review]
😅

**Comment by smoya** [comment]
> 😅

😝 Fixed

**Review by magicmatatjahu — APPROVED** [review]
For me it's ok, but lets see what codeowners will say :)

**Comment by smoya** [comment]
@derberg @M3lkior @fmvilas as code owners, would any of you mind doing a review to this? @dalelane already approved but more 👀  are needed. Thanks!

**Review by derberg — COMMENTED** [review]
can you please clarify what is the release strategy for these schema parsers for parser 2.0 ?

this PR points to `next-major` and 2.0 release candidate will be generated. Then the RC will be used in the `generator`?

**Comment by smoya** [comment]
still relevant

**Comment by M3lkior** [comment]
> can you please clarify what is the release strategy for these schema parsers for parser 2.0 ?
> 
> this PR points to `next-major` and 2.0 release candidate will be generated. Then the RC will be used in the `generator`?

there is a WIP PR to use it from the generator: https://github.com/asyncapi/generator/pull/833

**Comment by smoya** [comment]
> Just a last rebase is needed from `next-major` @smoya :)

@M3lkior Code is now up-to-date now. I also updated the `parser-js` package to the latest 2.0 prerelease.
